src/librustc_mir/interpret/memory.rs

```diff
@@ -808,8 +813,26 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
 
         // This checks relocation edges on the src.
         let src_bytes = self.get(src.alloc_id)?
-            .get_bytes_with_undef_and_ptr(&tcx, src, size)?
-            .as_ptr();
+            .get_bytes_with_undef_and_ptr(&tcx, src, size)?;
+
+        let src_bytes = match src_bytes {
+            BytesOrElidedUndef::MaybeBytes(bytes) => bytes.as_ptr(),
+            BytesOrElidedUndef::ElidedUndef => {
+                // All bytes are undef. There are no source bytes to copy from but if they were
+                // allocated right now they might be all zeroes. If the destination is not itself
+                // completely undef (which could also happen from `src == dest` then set all bytes
+                // in the range to undef and do nothing else.
```

Comment:
Where are the zeroes here coming from? This is just implementing a copy of a lot of uninitialized memory. Nothing weird. The important bit is that we do not call `get_bytes_mut` to avoid "allocating" the target allocation if it is not already allocated.

Also there is an unclosed parenthesis in your comment.

Reply:
It was a note that without eliding the undef allocation the implementatio is currently copying a bunch of zeroes—as the memory is currently allocated with `vec![0; len]`—and if anything breaks off of this since it from making the (wrong) assumption of getting zeroed bytes, this would explain the different state in memory.

> Also there is an unclosed parenthesis in your comment.

Would it calm your mind if I were to close it here ) :wink:

Reply:
Oh but we could use any other default value -- the value does not matter when the "initialized" mask is not set.

So this should emphasise on needlessly allocating in the target allocation, instead of needlessly copying. *That* is the main point here.